static_frame/core/index_hierarchy.py

```diff
@@ -1018,259 +1362,580 @@ def values_at_depth(self,
+
+            new_index: tp.Dict[tp.Hashable, int] = {}
+            index_remap: tp.Dict[tp.Hashable, tp.Hashable] = {}
+
+            for label_idx, label in enumerate(index.values):
+                new_label = get_new_label(label)
+
+                if new_label not in new_index:
+                    new_index[new_label] = len(new_index)
+                else:
+                    index_remap[label_idx] = new_index[new_label]
+
+            new_indices[level] = index.__class__(new_index)
+
+            indexer = np.array([index_remap.get(i, i) for i in self._indexers[level]])
+            new_indexers[level] = indexer
+
+        new_indexers.flags.writeable = False
+
+        return self.__class__(
+                indices=new_indices,
+                indexers=new_indexers,
+                name=self._name,
+                )
+
+    def rehierarch(self: IH,
+            depth_map: tp.Sequence[int],
+            ) -> IH:
+        '''
+        Return a new :obj:`IndexHierarchy` that conforms to the new depth assignments given be `depth_map`.
+        '''
+        if self._recache:
+            self._update_array_cache() # pragma: no cover
+
+        rehierarched_blocks, _ = rehierarch_from_type_blocks(
+                labels=self._blocks,
+                depth_map=depth_map,
+                )
+        return self.__class__._from_type_blocks(
+            blocks=rehierarched_blocks,
+            index_constructors=self._index_constructors,
+            own_blocks=True,
+            )
+
+    def _get_unique_labels_in_occurence_order(self: IH,
+            depth: int = 0,
+            ) -> tp.Sequence[tp.Hashable]:
+        '''
+        Index could be [A, B, C]
+        Indexers could be [2, 0, 0, 2, 1]
+
+        This function return [C, A, B] -- shoutout to my initials
+        '''
+        # get the outer level, or just the unique frame labels needed
+        labels = self.values_at_depth(depth)
+        label_indexes = ufunc_unique1d_positions(labels)[0]
+        label_indexes.sort()
+        return labels[label_indexes] # type: ignore
+
+    # --------------------------------------------------------------------------
+
+    def _build_mask_for_key_at_depth(self: IH,
+            depth: int,
+            key: tp.Union[np.ndarray, CompountLabelType],
+            ) -> np.ndarray:
+        '''
+        Determines the indexer mask for `key` at `depth`.
+        '''
+        # This private internal method assumes recache has already been checked for!
+
+        key_at_depth = key[depth]
+
+        # Key is already a mask!
+        if key_at_depth.__class__ is np.ndarray and key_at_depth.dtype == DTYPE_BOOL: # type: ignore
+            return key_at_depth
+
+        index_at_depth = self._indices[depth]
+        indexer_at_depth = self._indexers[depth]
+
+        if isinstance(key_at_depth, slice):
+            if key_at_depth.start is not None:
+                start: int = tp.cast(int, index_at_depth.loc_to_iloc(key_at_depth.start))
```

Comment:
Again, I am cautious in using `tp.cast` in anything that might be called very frequently.

Reply:
Okay I'll `# type: ignore`